refactor(ast)!: Change order of fields in CallExpression

cblh opened the pull request:
fix: #4821

cblh committed:
chore(ast): Change order of fields in CallExpression

cblh committed:
chore(ast): Change order of fields in CallExpression

cblh commented:
This test case failed. https://github.com/oxc-project/oxc/blob/7a75e0f8a77006ef32fa0db45a329d961e70bdf3/crates/oxc_linter/src/rules/eslint/no_this_before_super.rs#L384-L387

Dunqing commented:
> This test case failed.
> 
> https://github.com/oxc-project/oxc/blob/7a75e0f8a77006ef32fa0db45a329d961e70bdf3/crates/oxc_linter/src/rules/eslint/no_this_before_super.rs#L384-L387

Seems as expected, try to fix it

cblh committed:
fix(linter): handle `this` or `super` in arguments for `no-this-before-super` rule

- Added `contains_this_or_super_in_args` method to check for `this` or `super` usage in function call arguments.
- Modified logic in `NoThisBeforeSuper` rule to ensure that `super()` is correctly identified as being called before any usage of `this`.
- Introduced `contains_this_or_super` helper method for deeper inspection of arguments within expressions.
- Updated `Semantic` handling to support the new checks.

rzvxa commented:
This PR should be merged with #4851 to minimize our minor version bumps. Ideally, we'd get #4188 in this version too(at least the initial implementation).

rzvxa commented on crates/oxc_linter/src/rules/eslint/no_this_before_super.rs:
```diff
@@ -246,6 +251,27 @@ impl NoThisBeforeSuper {
         })
     }
+
+    fn contains_this_or_super(arg: &Argument, semantic: &Rc<Semantic>) -> bool {
```
You are not using this `semantic` parameter, I think you can safely remove it.

cblh replied:
The method `Self::contains_this_or_super_in_args(&call_expr.arguments, semantic)` is used to recursively handle cases like `super(a(b(this.c)))`.

rzvxa replied:
You are absolutely right, Semantic is needed to be passed around, however, that function only uses its `semantic` to call `Self::contains_this_or_super`:

```rust
    fn contains_this_or_super_in_args(args: &[Argument], semantic: &Rc<Semantic>) -> bool {
        args.iter().any(|arg| Self::contains_this_or_super(arg, semantic))
    }
```

And `Self::contains_this_or_super` only uses it to call `Self::contains_this_or_super_in_args`.
I think you've gotten rid of the need for semantic usage however the compiler because of cyclic recursion doesn't notice this to provide a dead_code warning.

Please let me know if I'm missing something.

cblh committed:
test(linter): add test case for `super(a(b(this.c)))` in `no-this-before-super` rule

- Added a new test case in the `no-this-before-super` rule to ensure it correctly disallows the usage of `this` or `super` in nested arguments like `super(a(b(this.c)))`.
- Updated the snapshot to reflect the new test case.

cblh committed:
refactor(linter): Remove unnecessary `Semantic` reference in `no_this_before_super` rule

- Removed the `Rc<Semantic>` reference from the `contains_this_or_super` and `contains_this_or_super_in_args` methods, simplifying their signatures.
- Removed the unused `Rc` import from `std::rc`.
- Updated method calls to reflect the removal of the `Semantic` argument.

These changes streamline the code by eliminating redundant dependencies, improving maintainability.

overlookmotel commented:
Fixed in #4851.

overlookmotel commented:
#4851 has already been merged, so I don't think we need to wait to merge this.

@cblh Could you please rebase on main to remove merge conflicts? (don't worry about fixing merge conflicts in `generated/derive_clone_in.rs` - just rebase then run `just ast` again to regenerate the file)

@rzvxa This currently shows as "change requested". If your requested changes have been made, please can you mark that as resolved?

cblh committed:
fix(oxc_ast, oxc_transformer): correct allocator parameter and argument positioning

- Updated `derive_clone_in.rs` to use the correct `allocator` parameter for consistency.
- Fixed argument positioning in function calls across `exponentiation_operator.rs` and `react/refresh.rs` to ensure proper parameter passing.
- These changes resolve potential runtime errors due to incorrect argument placement.

rzvxa commented:
> #4851 has already been merged, so I don't think we need to wait to merge this.
> 
> @cblh Could you please rebase on main to remove merge conflicts? (don't worry about fixing merge conflicts in `generated/derive_clone_in.rs` - just rebase then run `just ast` again to regenerate the file)
> 
> @rzvxa This currently shows as "change requested". If your requested changes have been made, please can you mark that as resolved?

That one proved not to contain any breaking changes for our downstream, But I guess we can do the same thing with this PR as well.

I'll test it against rolldown and let you know. If it doesn't break anything there we may drop the bang in the title. I'll take care of merging it when CI is green once more.

@cblh We had some naming refactors recently, You should be able to make the CI green by running `just ast` and committing the changes.

cblh commented:
> > #4851 has already been merged, so I don't think we need to wait to merge this.
> > @cblh Could you please rebase on main to remove merge conflicts? (don't worry about fixing merge conflicts in `generated/derive_clone_in.rs` - just rebase then run `just ast` again to regenerate the file)
> > @rzvxa This currently shows as "change requested". If your requested changes have been made, please can you mark that as resolved?
> 
> That one proved not to contain any breaking changes for our downstream, But I guess we can do the same thing with this PR as well.
> 
> I'll test it against rolldown and let you know. If it doesn't break anything there we may drop the bang in the title. I'll take care of merging it when CI is green once more.
> 
> @cblh We had some naming refactors recently, You should be able to make the CI green by running `just ast` and committing the changes.

done

rzvxa commented:
Thank you❤️, I'll take care of the rest.

rzvxa commented:
@Boshen This PR contains a few breaking changes in the [rolldown/ast_snippet.rs](https://github.com/rolldown/rolldown/blob/main/crates/rolldown_ecmascript/src/ast_snippet.rs).

I'll let you decide on merging it now or waiting to add it with other breaking changes to avoid bumping the minor version multiple times.

It contains a few linter updates so merging would be more desirable to avoid conflicts, However, most of the diff is from the generated code and snapshots. Resolving conflicts here shouldn't be that hard.

Dunqing committed:
Merge remote-tracking branch 'origin/main' into fix/4821

Dunqing committed:
Fix

Dunqing committed:
update semantic snap

Dunqing commented:
This change may affect the development of the transformer so I would like to merge this ASAP
